### PR TITLE
Migrate application targets to SPM

### LINKED ISSUE #2819: Crash playing reroute sound (SPM only) — RouteVoiceController.init(navigationService:speechSynthesizer:accessToken:host:)
As of #2808, if MapboxNavigation has been installed using SPM, it crashes trying to play the reroute sound when the user goes off-route:

```
Thread 1 Queue : com.apple.main-thread (serial)
#0	0x000000010348ba80 in _swift_runtime_on_report ()
#1	0x00000001035060e3 in _swift_stdlib_reportFatalErrorInFile ()
#2	0x0000000103195f30 in closure #1 in closure #1 in closure #1 in _assertionFailure(_:_:file:line:flags:) ()
#3	0x0000000103195a87 in closure #1 in closure #1 in _assertionFailure(_:_:file:line:flags:) ()
#4	0x0000000103195763 in closure #1 in _assertionFailure(_:_:file:line:flags:) ()
#5	0x00000001031953e1 in _assertionFailure(_:_:file:line:flags:) ()
#6	0x000000010052e1d5 in RouteVoiceController.init(navigationService:speechSynthesizer:accessToken:host:) at /path/to/mapbox-navigation-ios/Sources/MapboxNavigation/RouteVoiceController.swift:73
#7	0x000000010052dc0c in RouteVoiceController.__allocating_init(navigationService:speechSynthesizer:accessToken:host:) ()
#8	0x00000001004e6032 in NavigationViewController.init(for:routeIndex:routeOptions:navigationOptions:) at /path/to/mapbox-navigation-ios/Sources/MapboxNavigation/NavigationViewController.swift:286
#9	0x00000001004e473a in NavigationViewController.__allocating_init(for:routeIndex:routeOptions:navigationOptions:) ()
#10	0x000000010033332f in ViewController.navigationViewController(navigationService:) at /path/to/mapbox-navigation-ios/Example/ViewController.swift:396
#11	0x000000010032c96f in ViewController.startBasicNavigation() at /path/to/mapbox-navigation-ios/Example/ViewController.swift:216
#12	0x000000010032b53f in closure #1 in ViewController.presentActionsAlertController() at /path/to/mapbox-navigation-ios/Example/ViewController.swift:168
```

https://github.com/mapbox/mapbox-navigation-ios/blob/99e1236b2fd04df1202c86710494e9b2b58bf375/Sources/MapboxNavigation/RouteVoiceController.swift#L73

The root cause seems to be that the Package.swift doesn’t list reroute-sound or any of the other assets in Assets.xcassets as bundle resources:

https://github.com/mapbox/mapbox-navigation-ios/blob/99e1236b2fd04df1202c86710494e9b2b58bf375/Package.swift#L50-L59

/cc @mapbox/navigation-ios

### PR DESCRIPTION
**Pull request by 1ec5**
Migrated the Example, Example-CarPlay, and Bench from Carthage-built frameworks to static frameworks built by SPM from the colocated package.

Fixed a bunch of crashes due to bundles not being found. Since SPM builds static frameworks, the usual `Bundle(for:)` trick doesn’t work.

Static linking also breaks `canImport` weak linking, just like with CocoaPods, so even an application that explicitly depends on MapboxGeocoder wouldn’t enable the conditionally linked MapboxGeocoder usage in MapboxNavigation to build. For now, this PR strongly links MapboxGeocoder to fix those build errors, but #2285 tracks removing the dependency altogether. Similarly, Bench doesn’t build because of static linking: #2820.

Fixes #2819. Depends on #2808.

/cc @mapbox/navigation-ios

### CODE REVIEW

File: Sources/MapboxCoreNavigation/BundleAdditions.swift [hunk 1]
```diff
@@ -33,7 +33,11 @@ extension Bundle {
      The Mapbox Core Navigation framework bundle.
      */
     public class var mapboxCoreNavigation: Bundle {
+        #if SWIFT_PACKAGE
+        return .module
+        #else
         return Bundle(for: RouteController.self)
+        #endif
```
Comment (1ec5) [inline Sources/MapboxCoreNavigation/BundleAdditions.swift:40]: This change is fine, but I’m a little concerned about `mapboxNavigationIfInstalled` below. I don’t think there’s anything like `Bundle.module` for discovering a _different_ module’s bundle.

Comment (MaximAlien) [reply Sources/MapboxCoreNavigation/BundleAdditions.swift:40]: I think I'm getting crash in the same place when starting active guidance navigation.

**Comment by 1ec5** [comment]
> Since SPM builds static frameworks

I tried setting `type: .dynamic` for MapboxNavigation in Package.swift, but it results in an empty framework with nothing but an Info.plist. 🤷‍♂️ If we could get SPM to generate a dynamic framework for MapboxNavigation, then we’d be much closer to getting Carthage to build MapboxNavigation despite its dependencies coming from SPM.

File: Sources/MapboxCoreNavigation/BundleAdditions.swift (hunk 1 above)
Comment (MaximAlien) [reply Sources/MapboxCoreNavigation/BundleAdditions.swift:40]: Looks like force unwrapping shouldn't be used when accessing Bundle.mapboxCoreNavigation. E.g. this one crashes as well: 
```
let sdkVersion: String = String(describing: Bundle.mapboxCoreNavigation.object(forInfoDictionaryKey: "CFBundleShortVersionString")!)
```

**Comment by 1ec5** [comment]
These error messages about the binary dependences are very prevalent; I don’t think I’ve ever opened the project without seeing them and having to clear package caches:

```
artifact of binary target 'MapboxAccounts' failed extraction: The operation couldn’t be completed. (TSCBasic.StringError error 1.)

artifact of binary target 'MapboxMaps' failed extraction: The operation couldn’t be completed. (TSCBasic.StringError error 1.)

artifact of binary target 'MapboxCommon' failed extraction: The operation couldn’t be completed. (TSCBasic.StringError error 1.)

artifact of binary target 'MapboxNavigationNative' failed extraction: The operation couldn’t be completed. (TSCBasic.StringError error 1.)

artifact of binary target 'MapboxCoreMaps' failed extraction: The operation couldn’t be completed. (TSCBasic.StringError error 1.)
```

**Comment by MaximAlien** [comment]
Overall looks good, just got additional crash because of force unwrapping in:
```
mobileEventsManager.initialize(withAccessToken: accessToken, userAgentBase: userAgent, hostSDKVersion: String(describing: Bundle.mapboxCoreNavigation.object(forInfoDictionaryKey: "CFBundleShortVersionString")!))
```

After removing `!` was able to run example successfully.

File: MapboxNavigation.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -2050,17 +1910,16 @@
 				C53F2EE320EBC95600D9798F /* Sources */,
 				C53F2EE920EBC95600D9798F /* Frameworks */,
 				C53F2EF220EBC95600D9798F /* Resources */,
-				C53F2EF920EBC95600D9798F /* Embed Frameworks */,
 				C53F2F0320EBC95600D9798F /* Apply Mapbox Access Token */,
-				DA3D112B25B0AB0100A6AB8C /* Copy Frameworks with Carthage */,
```
Comment (MaximAlien) [inline MapboxNavigation.xcodeproj/project.pbxproj:2055]: Shalle we remove `input.xcfilelist` and `output.xcfilelist` as well?

**Review by MaximAlien — APPROVED** [review]

**Comment by MaximAlien** [comment]
`Bench` target cannot be built with errors like:
```
Swift package product 'MapboxGeocoder' is linked as a static library by 'Bench' and 'TestHelper'. This will result in duplication of library code.
```

**Comment by 1ec5** [comment]
Yes, unfortunately, Bench will fail until #2820 is fixed.

**Pull request merged**